File: EmojibleIOS/EmojibleIOS/Models/ModelFactories/UserFactory.swift
//
//  UserFactory.swift
//  EmojibleIOS
//
//  Created by Haldun Balim on 12.11.2020.
//

import Foundation

class UserFactory{
    
    func create(userType:String, email:String, name:String, surname:String, birthDate:Date, classIds:[String]=[], classId:String?=nil)->UserModel{
        if userType == "Teacher"{
            return TeacherModel(email: email, name: name, surname: surname, birthDate: birthDate,classIds: classIds)
        }else{
            return StudentModel(email:email, name:name, surname:surname, birthDate:birthDate, classId: classId)
        }
    }
    
    func create(dict: [String:Any]) -> UserModel{
        let userType = dict["userType"] as! String
        if userType == "Student"{
            return StudentModel(dictionary: dict)
        }else{
            return TeacherModel(dictionary: dict)
        }
    }
    
    private init(){}
    private static let instance = UserFactory()
    public static func getInstance() -> UserFactory{
        return .instance
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/models/modelFactories/UserFactory.kt
package com.dji.emojibleandroid.models.modelFactories

import com.dji.emojibleandroid.models.StudentModel
import com.dji.emojibleandroid.models.TeacherModel
import com.dji.emojibleandroid.models.UserModel


class UserFactory private constructor(){

    fun create(userType: String, email: String, name: String, surname: String, birthDate: String, classIds: MutableList<String> = mutableListOf(), classId: String? = null): UserModel{
        return when(userType) {
            "Teacher" -> TeacherModel(email, name, surname, birthDate, classIds)
            "Student" -> StudentModel(email, name, surname, birthDate, classId)
            else -> throw Exception("userType $userType is not a valid userType")
        }
    }

    fun create(dictionary: HashMap<String, Any>): UserModel {
        val userType: String = dictionary["userType"] as String
        return when (userType) {
            "Teacher" -> TeacherModel(dictionary)
            "Student" -> StudentModel(dictionary)
            else -> throw Exception("userType $userType is not a valid userType")
        }
    }

    private object HOLDER {
        val INSTANCE = UserFactory()
    }

    companion object {
        val instance: UserFactory by lazy { HOLDER.INSTANCE }
    }
}